### PR TITLE
Extend MemoryCacheOptions with TrackLinkedCacheEntries

### LINKED ISSUE #45592: [MemoryCache] Add possibility to disable linked cache entries
## Background and Motivation

`MemoryCache` by default tracks linked cache entries to allow for propagating options. 

In the following example, the `expirationToken` added for `child` is also applied to `parent`:

```cs
using (var parent = cache.CreateEntry(key))
{
    parent.SetValue(obj);

    using (var child = cache.CreateEntry(key1))
    {
        child.SetValue(obj);
        child.AddExpirationToken(expirationToken);
    }
}
```

Tracking internally uses `AsyncLocal<T>`, which is expensive and adds non-trivial overhead. The cost of using it has popped out in the performance reports that we have received from our customers (#45436).

Currently, this feature can not be disabled. 

## Proposed API

As suggested by @Tratcher in https://github.com/dotnet/runtime/issues/45436#issuecomment-737345755, we can extend the existing `MemoryCacheOptions` with a flag that disables this behaviour.

Naming is hard and the best idea I currently have is `TrackLinkedCacheEntries`:

```diff
namespace Microsoft.Extensions.Caching.Memory
{
    public class MemoryCacheOptions
    {
+        public bool TrackLinkedCacheEntries { get ; init; } = true;
    }
}
```

I am open to suggestions.

## Usage Examples

```cs
var options = new MemoryCacheOptions
{
    TrackLinkedCacheEntries = false
};

var cache = new MemoryCache(options);
```

## Alternative Designs

As an alternative, we could add a new method to `MemoryCache` that would allow for creating cache entries without tracking:

```diff
namespace Microsoft.Extensions.Caching.Memory
{
    public class MemoryCache
    {
        public ICacheEntry CreateEntry(object key); // existing method
+        public ICacheEntry CreateUnlinkedEntry(object key);
    }
```

But there is a lot of existing extension methods that allow for adding new cache entries to the cache and we would most probably need to add new overloads for them...

## Risks

Introducing this API has no risks as long as we don't change the default settings (enabled by default).

cc @eerhardt @maryamariyan @davidfowl

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #45592

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs [hunk 1]
```diff
@@ -53,6 +53,12 @@ public double CompactionPercentage
             }
         }
 
+        /// <summary>
+        /// Gets or sets the flat that controls linked entries tracking. Disabled by default.
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:57]:
```suggestion
        /// Gets or sets whether to track linked entries. Disabled by default.
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs [hunk 2]
```diff
@@ -53,6 +53,12 @@ public double CompactionPercentage
...
+        /// </summary>
+        /// <remarks>Prior to .NET 7 this feature was always enabled.</remarks>
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:59]: Does that mean this is a breaking change?  Above it says it's disabled by default but here it says it was previously always enabled.

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:59]: > Does that mean this is a breaking change? Above it says it's disabled by default but here it says it was previously always enabled.

Yes, exactly. I've added `needs-breaking-change-doc-created` to ensure we document that.

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCacheOptions.cs:59]: This has been discussed and approved at the API review: https://github.com/dotnet/runtime/issues/45592#issuecomment-896214586

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs [hunk 3]
```diff
@@ -232,7 +235,7 @@ public bool TryGetValue(object key, out object result)
                     entry.LastAccessed = utcNow;
                     result = entry.Value;
 
-                    if (entry.CanPropagateOptions())
+                    if (_options.TrackLinkedCacheEntries && entry.CanPropagateOptions())
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:238]: Should use the "stored setting value", so it is consistent for the entire MemoryCache lifetime. 

```suggestion
                    if (TrackLinkedCacheEntries && entry.CanPropagateOptions())
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs [hunk 4]
```diff
@@ -28,7 +28,7 @@ internal CacheEntry(object key, MemoryCache memoryCache)
         {
             Key = key ?? throw new ArgumentNullException(nameof(key));
             _cache = memoryCache ?? throw new ArgumentNullException(nameof(memoryCache));
-            _previous = CacheEntryHelper.EnterScope(this);
+            _previous = memoryCache.TrackLinkedCacheEntries ? CacheEntryHelper.EnterScope(this) : null;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs:31]: Need to update the comment above on `_previous`'s declaration:

> private CacheEntry _previous; // this field is not null only before the entry is added to the cache

That comment is now wrong.

**Comment by eerhardt** [comment]
>         public async Task OnceExpiredIsSetToTrueItRemainsTrue()

Any reason why this test wasn't turned into a Theory?

---
Refers to: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CacheEntryScopeExpirationTests.cs:513 in 4684c51. [](commit_id = 4684c5185ecf90aa19390d375871cdef6820f3cf, deletion_comment = False)

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs (hunk 4 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs:31]: good catch

**Comment by adamsitnik** [comment]
> Any reason why this test wasn't turned into a Theory?

There is just one entry being created in this test (no dependencies to other linked entries) and it just verifies whether setting entry to expired is thread-safe (not related to tracking)

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Comment by eerhardt** [comment]
Failures are unrelated. Timing out due to dotnet/core-eng#14346

**Comment by EgorBo** [comment]
arm64 improvements: https://github.com/dotnet/perf-autofiling-issues/issues/1537

**Comment by sebastienros** [comment]
@adamsitnik someone finally screamed

**Comment by ericstj** [comment]
@adamsitnik were you planning to create a breaking change document for this?

**Comment by maryamariyan** [comment]
I just added https://github.com/dotnet/docs/issues/31739

**Comment by adamsitnik** [comment]
> I just added https://github.com/dotnet/docs/issues/31739

@maryamariyan big thanks for doing that!

**Pull request merged**